### PR TITLE
Improve logging

### PR DESCRIPTION
**Pull request by tobim**
This PR attempts to make the VASTs log more consistent.

Tasks:
- [x] Let `VAST_{ERROR,WARNING,INFO,DEBUG}` reject `const char*` as its first parameter
- [x] Add `VAST_{ERROR,WARNING,INFO,DEBUG}_` to bypass this check.
- [x] Convert messages to *subject verb object* style
- [x] Reserve `VAST_INFO` to the main system actors (SINK, SOURCE, IMPORTER, ARCHIVE, INDEX, EXPORTER)
- [x] Add logging guidelines to`CONTRIBUTING.md`

### CODE REVIEW

File: libvast/src/system/node_command.cpp [hunk 1]
```diff
@@ -99,7 +99,7 @@ node_command::connect_to_node(scoped_actor& self,
     err += endpoint_str;
     return make_error(sec::invalid_argument, std::move(err));
   }
-  VAST_INFO(&self, "connect to remote node:", id);
+  VAST_DEBUG(&self, "connect to remote node:", id);
```
Comment (mavam) [inline libvast/src/system/node_command.cpp]: s/connect/connects/

**Comment by mavam** [comment]
Perhaps squash some of the commits, @tobim? Seems like there are some repetitions.

File: CMakeLists.txt [hunk 2]
```diff
@@ -37,6 +37,10 @@ list(GET version_numbers 0 VERSION_MAJOR)
 list(GET version_numbers 1 VERSION_MINOR)
 set(VERSION_MAJ_MIN "${VERSION_MAJOR}.${VERSION_MINOR}")
 
+if (NOT VAST_LOG_LEVEL)
+  set(VAST_LOG_LEVEL "-1")
```
Comment (mavam) [inline CMakeLists.txt]: Would you mind elaborating why you set the log level to a string an not a numeric value?

File: libvast/src/column_index.cpp [hunk 3]
```diff
@@ -95,20 +95,20 @@ caf::error column_index::init() {
   if (exists(filename_)) {
     detail::value_index_inspect_helper tmp{index_type_, idx_};
     if (auto err = load(sys_, filename_, last_flush_, tmp)) {
-      VAST_ERROR("unable to load value index from disk", sys_.render(err));
+      VAST_ERROR(this, "unable to load value index from disk", sys_.render(err));
```
Comment (mavam) [inline libvast/src/column_index.cpp]: Seems like the sentence doesn't quite read properly: `this, "failed to load..."`?

File: libvast/src/column_index.cpp [hunk 4]
```diff
@@ -118,7 +118,7 @@ caf::error column_index::flush_to_disk() {
   auto offset = idx_->offset();
   if (offset == last_flush_)
     return caf::none;
-  VAST_DEBUG("flush index (" << (offset - last_flush_) << '/' << offset,
+  VAST_DEBUG(this, "flush index (" << (offset - last_flush_) << '/' << offset,
```
Comment (mavam) [inline libvast/src/column_index.cpp]: `flush` should read `flushes` or `flushed`.

File: libvast/src/meta_index.cpp [hunk 5]
```diff
@@ -49,11 +49,12 @@ operator[](const uuid& partition) const {
 void meta_index::add(const uuid& partition,
                      const const_table_slice_handle& slice) {
   auto& rng = partitions_[partition].range;
+  VAST_DEBUG(this, "meta_index got table slice:", slice);
```
Comment (mavam) [inline libvast/src/meta_index.cpp]: This doesn't read like a proper sentence.

File: libvast/src/segment_store.cpp [hunk 6]
```diff
@@ -57,7 +57,7 @@ segment_store::~segment_store() {
 
 caf::error segment_store::put(const_table_slice_handle xs) {
   VAST_TRACE(VAST_ARG(xs));
-  VAST_DEBUG("adding a table slice");
+  VAST_DEBUG(this, "adding a table slice");
```
Comment (mavam) [inline libvast/src/segment_store.cpp]: `adds`?

File: libvast/src/segment_store.cpp [hunk 7]
```diff
@@ -78,8 +78,8 @@ caf::error segment_store::flush() {
     return err;
   // Keep new segment in the cache.
   cache_.emplace(seg_ptr->id(), seg_ptr);
-  VAST_DEBUG("wrote new segment to", filename.trim(-3));
-  VAST_DEBUG("saving segment meta data");
+  VAST_DEBUG(this, "wrote new segment to", filename.trim(-3));
+  VAST_DEBUG(this, "saving segment meta data");
```
Comment (mavam) [inline libvast/src/segment_store.cpp]: `saves`?

File: libvast/src/segment_store.cpp [hunk 8]
```diff
@@ -88,7 +88,7 @@ segment_store::get(const ids& xs) {
   VAST_TRACE(VAST_ARG(xs));
   // Collect candidate segments by seeking through the ID set and
   // probing each ID interval.
-  VAST_DEBUG("getting table slices with ids");
+  VAST_DEBUG(this, "getting table slices with ids");
```
Comment (mavam) [inline libvast/src/segment_store.cpp]: s/getting/retrieves/ ?

File: libvast/src/segment_store.cpp [hunk 9]
```diff
@@ -103,24 +103,24 @@ segment_store::get(const ids& xs) {
     return error;
   // Process candidates in reverse order for maximum LRU cache hits.
   std::vector<const_table_slice_handle> result;
-  VAST_DEBUG("processing", candidates.size(), "candidates");
+  VAST_DEBUG(this, "processing", candidates.size(), "candidates");
```
Comment (mavam) [inline libvast/src/segment_store.cpp]: `processes`

File: libvast/src/segment_store.cpp [hunk 10]
```diff
@@ -103,24 +103,24 @@ segment_store::get(const ids& xs) {
...
   for (auto cand = candidates.rbegin(); cand != candidates.rend(); ++cand) {
     auto& id = *cand;
     caf::expected<std::vector<const_table_slice_handle>> slices{caf::no_error};
     if (id == builder_.id()) {
-      VAST_DEBUG("looking into builder");
+      VAST_DEBUG(this, "looking into builder");
```
Comment (mavam) [inline libvast/src/segment_store.cpp]: `looks`

File: libvast/src/segment_store.cpp [hunk 11]
```diff
@@ -103,24 +103,24 @@ segment_store::get(const ids& xs) {
...
-        VAST_DEBUG("got cache hit for segment", id);
+        VAST_DEBUG(this, "got cache hit for segment", id);
         seg_ptr = i->second;
       } else {
-        VAST_DEBUG("got cache miss for segment", id);
+        VAST_DEBUG(this, "got cache miss for segment", id);
         auto fname = segment_path() / to_string(id);
         if (auto err = load(sys_, fname, seg_ptr)) {
-          VAST_ERROR("unable to load segment:", sys_.render(err));
+          VAST_ERROR(this, "unable to load segment:", sys_.render(err));
```
Comment (mavam) [inline libvast/src/segment_store.cpp]: `failed to load...`

File: libvast/src/system/exporter.cpp [hunk 12]
```diff
@@ -42,9 +42,11 @@ namespace system {
 namespace {
 
 void ship_results(stateful_actor<exporter_state>* self) {
-  if (self->state.results.empty() || self->state.stats.requested == 0)
+  VAST_TRACE("");
```
Comment (mavam) [inline libvast/src/system/exporter.cpp]: What's the benefit of an empty string as trace log entry?

File: libvast/src/system/node.cpp [hunk 13]
```diff
@@ -306,7 +306,7 @@ caf::behavior node(node_ptr self, std::string id, path dir) {
         send(self, args);
       } else {
         auto e = make_error(ec::unspecified, "invalid command", cmd);
-        VAST_INFO(self, self->system().render(e));
+        VAST_WARNING(self, self->system().render(e));
```
Comment (mavam) [inline libvast/src/system/node.cpp]: Maybe add a little message here, like `failed to parse command`.

File: libvast/src/system/node_command.cpp [hunk 14]
```diff
@@ -99,7 +99,7 @@ node_command::connect_to_node(scoped_actor& self,
     err += endpoint_str;
     return make_error(sec::invalid_argument, std::move(err));
   }
-  VAST_INFO("connect to remote node:", id);
+  VAST_DEBUG(&self, "connect to remote node:", id);
```
Comment (mavam) [inline libvast/src/system/node_command.cpp]: `connects`

File: libvast/src/system/node_command.cpp [hunk 15]
```diff
@@ -109,7 +109,7 @@ node_command::connect_to_node(scoped_actor& self,
   auto host = node_endpoint.host;
   if (node_endpoint.host.empty())
     node_endpoint.host = "127.0.0.1";
-  VAST_INFO("connecting to", node_endpoint.host << ':' << node_endpoint.port);
+  VAST_INFO(&self, "connecting to", node_endpoint.host << ':' << node_endpoint.port);
```
Comment (mavam) [inline libvast/src/system/node_command.cpp]: `connects`

File: libvast/src/system/partition.cpp [hunk 16]
```diff
@@ -46,7 +46,7 @@ partition::partition(caf::actor_system& sys, const path& base_dir, uuid id,
   // are pre-loading all INDEXER types we are aware of.
   if (exists(dir_)) {
     if (auto err = load(sys_, meta_file(), meta_data_)) {
-      VAST_ERROR("unable to read partition meta data:", sys_.render(err));
+      VAST_ERROR(this, "unable to read meta data:", sys_.render(err));
```
Comment (mavam) [inline libvast/src/system/partition.cpp]: s/unable/failed/

File: libvast/src/system/reader_command_base.cpp [hunk 17]
```diff
@@ -69,20 +69,20 @@ int reader_command_base::run_impl(caf::actor_system& sys,
...
-        VAST_ERROR("no importers available at node", id);
+        VAST_ERROR(this, "did not receive any importers from node", id);
         stop = true;
       } else if (reg.components[id].count("importer") > 1) {
-        VAST_ERROR("multiple IMPOTER actors are not yet supported");
+        VAST_ERROR(this, "does not support multiple IMPORTER actors yet");
         stop = true;
       } else {
-        VAST_DEBUG("connecting source to importer");
+        VAST_DEBUG(this, "connecting to importer");
```
Comment (mavam) [inline libvast/src/system/reader_command_base.cpp]: `connects`

File: libvast/src/system/writer_command_base.cpp [hunk 18]
```diff
@@ -47,7 +47,7 @@ int writer_command_base::run_impl(caf::actor_system& sys,
     self->send_exit(sig_mon, exit_reason::user_shutdown);
   });
   // Spawn a sink.
-  VAST_DEBUG("spawning sink with parameters:", deep_to_string(options));
+  VAST_DEBUG(this, "spawning sink with parameters:", deep_to_string(options));
```
Comment (mavam) [inline libvast/src/system/writer_command_base.cpp]: `spawns`

File: libvast/src/system/writer_command_base.cpp [hunk 19]
```diff
@@ -70,14 +70,14 @@ int writer_command_base::run_impl(caf::actor_system& sys,
     args += make_message("--unified");
   auto max_events = get_or<uint64_t>(options, "events", 0u);
   args += make_message("-e", std::to_string(max_events));
-  VAST_DEBUG("spawning exporter with parameters:", to_string(args));
+  VAST_DEBUG(this, "spawning exporter with parameters:", to_string(args));
```
Comment (mavam) [inline libvast/src/system/writer_command_base.cpp]: `spawns`

File: libvast/src/table_index.cpp [hunk 20]
```diff
@@ -261,7 +261,7 @@ caf::expected<bitmap> table_index::lookup_impl(const predicate& pred,
     // Redirect to ordinary data lookup on column 0.
     return lookup_impl(pred, dx, x);
   }
-  VAST_WARNING("unsupported attribute:", ex.attr);
+  VAST_WARNING(this, "unsupported attribute:", ex.attr);
```
Comment (mavam) [inline libvast/src/table_index.cpp]: `got unsupported...`

File: libvast/src/table_index.cpp [hunk 21]
```diff
@@ -280,7 +280,7 @@ caf::expected<bitmap> table_index::lookup_impl(const predicate& pred,
   VAST_ASSERT(t);
   auto index = r.flat_index_at(dx.offset);
   if (!index) {
-    VAST_DEBUG("invalid offset for record type", dx.type);
+    VAST_DEBUG(this, "invalid offset for record type", dx.type);
```
Comment (mavam) [inline libvast/src/table_index.cpp]: `got invalid offset...`

File: libvast/src/table_slice.cpp [hunk 22]
```diff
@@ -76,7 +76,8 @@ table_slice_ptr table_slice::make_ptr(record_type layout,
   using factory_fun = table_slice_ptr (*)(record_type);
   auto val = sys.runtime_settings().get(impl);
   if (!caf::holds_alternative<generic_fun>(val)) {
-    VAST_ERROR("no factory function stored for implementation key", impl);
+    VAST_ERROR_("table_slice",
+                  "has no factory function for implementation key", impl);
```
Comment (mavam) [inline libvast/src/table_slice.cpp]: Style: indentation.

File: libvast/vast/logger.hpp [hunk 23]
```diff
@@ -48,34 +50,93 @@
...
 #define VAST_LOG(...) VAST_PP_OVERLOAD(VAST_LOG_, __VA_ARGS__)(__VA_ARGS__)
 
-#if CAF_LOG_LEVEL >= CAF_LOG_LEVEL_TRACE
+namespace vast::detail {
+template <class T>
+auto id_or_name(T&& x) {
+  static_assert(
+    !std::is_same_v<const char*, std::remove_reference<T>>,
+    "const char* is not allowed for the first argument in a logging statement"
+    "supply a component or use `VAST_[ERROR|WARNING|INFO|DEBUG]_ANON` instead");
```
Comment (mavam) [inline libvast/vast/logger.hpp]: Is `_ANON` now simply a `_` suffix?

File: libvast/vast/logger.hpp [hunk 24]
```diff
@@ -48,34 +50,93 @@
...
+
+  using Type = std::remove_pointer_t<T>;
```
Comment (mavam) [inline libvast/vast/logger.hpp]: Capitalized type aliases conflict with our coding style.

File: CMakeLists.txt (hunk 2 above)
Comment (tobim) [reply CMakeLists.txt]: cmake does not have an integral type. There is no semantic difference between `-1` and `"-1"`. I added the quotation marks to not confuse the reader about this.

File: libvast/vast/logger.hpp (hunk 23 above)
Comment (tobim) [reply libvast/vast/logger.hpp]: Yes, I missed updating that error message.
The suffix transitioned from `_FF` to `_ANON` to `_NS` (for *no subject*). But all of them suggest a wrong meaning at some use-sites. Maybe `_CS` for *custom subject*?

File: libvast/src/system/node_command.cpp (hunk 1 above)
Comment (mavam) [reply libvast/src/system/node_command.cpp]: Also, `self` should work here as well without the ampersand.

File: libvast/vast/logger.hpp (hunk 23 above)
Comment (tobim) [reply libvast/vast/logger.hpp]: As per our discussion, reverted to `_ANON`.

File: libvast/src/meta_index.cpp (hunk 5 above)
Comment (tobim) [reply libvast/src/meta_index.cpp]: done

**Comment by tobim** [comment]
I think all change requests are now addressed in 03878594a0c333766a422bcfec07a3f1b671ddd2.

**Comment by mavam** [comment]
I still see several unaddressed comments, e.g., `...(this, "getting table slice...`.

Per my earlier request, would you mind squashing the commits a bit? There are two identically named commits in there, which seems like they should be one commit.

File: libvast/src/system/exporter.cpp (hunk 12 above)
Comment (tobim) [reply libvast/src/system/exporter.cpp]: You are right to point out that messages like this are a net negative if the log format is not chosen appropriately. The current console logger will just show `ENTER` and `EXIT` messages. But the file logger writes the filename, line and function name, which is mostly what we want from the traces. Also, there is some precedent in the code base for doing it that way.

I would like to extend the logger to support different levels for file and console output. Then the empty trace messages won't be a problem any more.

**Comment by tobim** [comment]
Thanks for pointing that out again, I updated this PR once more and hope that I did not miss any open requests this time. Also, I squashed the two "Tune log levels" commits into one.

File: libvast/src/segment_store.cpp (hunk 11 above)
Comment (mavam) [reply libvast/src/segment_store.cpp]: What about this one, @tobim?

(By the way, github has now a "resolved" feature where you can check off each addressed review comment.)

File: libvast/src/system/exporter.cpp (hunk 12 above)
Comment (mavam) [reply libvast/src/system/exporter.cpp]: > Also, there is some precedent in the code base for doing it that way.

@Neverlord, would you mind filling in the rationale for those? I think these come from you.

> I would like to extend the logger to support different levels for file and console output. Then the empty trace messages won't be a problem any more.

Makes sense. That seems to be a straight shot in the CAF codebase.

Comment (Neverlord) [reply libvast/src/system/exporter.cpp]: This emits valuable information for following program flow with ENTER and EXIT messages for the function. This is used whenever no function arguments are available (other than `self` or the implicit `this` pointer) or are omitted for some other reason (not reasonably printable or something along those lines).

**Comment by mavam** [comment]
Great work, this is the foundation for more usability and easier debugging in the future. Can't wait to sprinkle the new macros over my code. 😉

**Pull request merged**